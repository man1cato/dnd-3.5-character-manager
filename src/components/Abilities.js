import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
