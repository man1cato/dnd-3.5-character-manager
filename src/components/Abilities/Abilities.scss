@import '../../styles/base/settings';

.Abilities {  
   display: grid;
   grid-template-columns: repeat(4 1fr);
   grid-auto-flow: dense;
   grid-row-gap: $xs-size; 
   text-align: center;
   align-items: center;
}