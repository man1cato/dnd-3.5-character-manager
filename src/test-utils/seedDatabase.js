import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}