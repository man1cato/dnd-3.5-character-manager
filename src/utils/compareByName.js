export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}