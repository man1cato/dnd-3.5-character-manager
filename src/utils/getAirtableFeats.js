import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}