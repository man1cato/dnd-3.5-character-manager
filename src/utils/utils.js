import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])