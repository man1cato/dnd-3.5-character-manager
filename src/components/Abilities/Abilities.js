import React, { Fragment, useState, useEffect } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import Counter from '../Counter/Counter'
import { calcAbilityMod, convertInputValue } from '../../utils/utils'
import './Abilities.scss'


const Abilities = ({ abilities, handleUpdate }) => {
	const scores = _.mapValues(abilities, (obj) => obj.score)
	const [tempScores, setTempScores] = useState(_.mapValues(abilities, (obj) => obj.tempScore))

	useEffect(() => {
		handleUpdate({ 
			abilities: _.mapValues(tempScores, (val, ability) => ({
				score: scores[ability],
				tempScore: val
			})) 
		})
	}, [tempScores])

	return (
		<div className="Abilities">
			<h5 className="grid__col1">Ability</h5>
			<h5 className="grid__col2">Score</h5>
			<h5 className="grid__col3">Temp Score</h5>
			<h5 className="grid__col4">Mod</h5>

			{_.map(['str', 'dex', 'con', 'int', 'wis', 'cha'], ability => {
				const score = scores[ability]
				const tempScore = tempScores[ability]
				return (
					<Fragment key={ability}>
						<div className="grid__col1">{ability.toUpperCase()}</div>
						<div className="grid__col2">{score}</div>
						<Counter
							className="grid__col3"
							value={convertInputValue(tempScore)}
							updateValue={value => setTempScores(update(tempScores, {
								[ability]: { $set: convertInputValue(value) }
							}))}
						/>
						<div 
							className="grid__col4"
							data-testid={ability + 'Mod'}
						>
							{!!tempScore ? calcAbilityMod(tempScore) : calcAbilityMod(score)}
						</div>
					</Fragment>
				)
			})}
		</div>
	)
}


export default Abilities