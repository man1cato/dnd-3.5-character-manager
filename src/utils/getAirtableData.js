import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}