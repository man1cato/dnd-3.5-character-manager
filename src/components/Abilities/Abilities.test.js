import React from 'react'
import { render, fireEvent, getNodeText } from '@testing-library/react'

import Abilities from './Abilities'
import { characterOne } from '../../tests/seedDatabase'
import { calcAbilityMod } from '../../utils/utils'


const handleUpdate = jest.fn()
let props
beforeEach(() => {
  props = {
    abilities: characterOne.abilities,
    handleUpdate
  }
})

test('should render Abilities with profile data', () => {
  const { container } = render(<Abilities {...props} />)
  expect(container.firstChild).toMatchSnapshot()
})

test('should update mod value when temp score input changed', () => {
  const { getAllByDisplayValue, getByTestId } = render(<Abilities {...props} />)
  const strInput = getAllByDisplayValue('0')[0]
  const strMod = getByTestId('strMod')
  const value = 20
  const mod = calcAbilityMod(value)

  fireEvent.change(strInput, { target: { value } })
  expect(getNodeText(strMod)).toBe(`${mod}`)
})