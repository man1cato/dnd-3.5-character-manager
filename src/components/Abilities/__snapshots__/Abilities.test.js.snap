// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Abilities with profile data 1`] = `
<Abilities
  abilities={
    Object {
      "cha": Object {
        "score": 11,
      },
      "con": Object {
        "score": 16,
      },
      "dex": Object {
        "score": 18,
      },
      "int": Object {
        "score": 20,
      },
      "str": Object {
        "score": 14,
      },
      "wis": Object {
        "score": 14,
      },
    }
  }
  handleUpdate={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "abilities": Object {
              "cha": Object {
                "score": 11,
                "tempScore": undefined,
              },
              "con": Object {
                "score": 16,
                "tempScore": undefined,
              },
              "dex": Object {
                "score": 18,
                "tempScore": undefined,
              },
              "int": Object {
                "score": 20,
                "tempScore": undefined,
              },
              "str": Object {
                "score": 14,
                "tempScore": undefined,
              },
              "wis": Object {
                "score": 14,
                "tempScore": undefined,
              },
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="Abilities"
  >
    <h5
      className="grid__col1"
    >
      Ability
    </h5>
    <h5
      className="grid__col2"
    >
      Score
    </h5>
    <h5
      className="grid__col3"
    >
      Temp Score
    </h5>
    <h5
      className="grid__col4"
    >
      Mod
    </h5>
    <div
      className="grid__col1"
    >
      STR
    </div>
    <div
      className="grid__col2"
    >
      14
    </div>
    <Counter
      className="grid__col3"
      updateValue={[Function]}
      value={0}
    >
      <div
        className="Counter grid__col3"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="remove"
            size="small"
          />
        </button>
        <input
          className="counter-value"
          onChange={[Function]}
          size={3}
          value={0}
        />
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="add"
            size="small"
          />
        </button>
      </div>
    </Counter>
    <div
      className="grid__col4"
    >
      2
    </div>
    <div
      className="grid__col1"
    >
      DEX
    </div>
    <div
      className="grid__col2"
    >
      18
    </div>
    <Counter
      className="grid__col3"
      updateValue={[Function]}
      value={0}
    >
      <div
        className="Counter grid__col3"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="remove"
            size="small"
          />
        </button>
        <input
          className="counter-value"
          onChange={[Function]}
          size={3}
          value={0}
        />
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="add"
            size="small"
          />
        </button>
      </div>
    </Counter>
    <div
      className="grid__col4"
    >
      4
    </div>
    <div
      className="grid__col1"
    >
      CON
    </div>
    <div
      className="grid__col2"
    >
      16
    </div>
    <Counter
      className="grid__col3"
      updateValue={[Function]}
      value={0}
    >
      <div
        className="Counter grid__col3"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="remove"
            size="small"
          />
        </button>
        <input
          className="counter-value"
          onChange={[Function]}
          size={3}
          value={0}
        />
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="add"
            size="small"
          />
        </button>
      </div>
    </Counter>
    <div
      className="grid__col4"
    >
      3
    </div>
    <div
      className="grid__col1"
    >
      INT
    </div>
    <div
      className="grid__col2"
    >
      20
    </div>
    <Counter
      className="grid__col3"
      updateValue={[Function]}
      value={0}
    >
      <div
        className="Counter grid__col3"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="remove"
            size="small"
          />
        </button>
        <input
          className="counter-value"
          onChange={[Function]}
          size={3}
          value={0}
        />
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="add"
            size="small"
          />
        </button>
      </div>
    </Counter>
    <div
      className="grid__col4"
    >
      5
    </div>
    <div
      className="grid__col1"
    >
      WIS
    </div>
    <div
      className="grid__col2"
    >
      14
    </div>
    <Counter
      className="grid__col3"
      updateValue={[Function]}
      value={0}
    >
      <div
        className="Counter grid__col3"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="remove"
            size="small"
          />
        </button>
        <input
          className="counter-value"
          onChange={[Function]}
          size={3}
          value={0}
        />
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="add"
            size="small"
          />
        </button>
      </div>
    </Counter>
    <div
      className="grid__col4"
    >
      2
    </div>
    <div
      className="grid__col1"
    >
      CHA
    </div>
    <div
      className="grid__col2"
    >
      11
    </div>
    <Counter
      className="grid__col3"
      updateValue={[Function]}
      value={0}
    >
      <div
        className="Counter grid__col3"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="remove"
            size="small"
          />
        </button>
        <input
          className="counter-value"
          onChange={[Function]}
          size={3}
          value={0}
        />
        <button
          onClick={[Function]}
          type="button"
        >
          <ion-icon
            name="add"
            size="small"
          />
        </button>
      </div>
    </Counter>
    <div
      className="grid__col4"
    >
      0
    </div>
  </div>
</Abilities>
`;
